dagutils/traversal.go

```diff
@@ -0,0 +1,384 @@
+package dagutils
+
+import (
+	"context"
+	"errors"
+
+	ipld "gx/ipfs/QmZtNq8dArGfnpCZfx2pUNY7UcjGhVp5qqwQ4hH6mpTMRQ/go-ipld-format"
+)
+
+// Traversal provides methods to move through a DAG of IPLD nodes
+// using an iterative algorithm and a handler for the user to perform
+// some logic on each visited node. This enables the implementation of
+// operations like reading and seeking inside of a UnixFS file DAG. In
+// contrast with a recursive solution, the implicit path in the call
+// stack is made explicit through the `path` variable.
+//
+// It exposes simple traversal operations like `Iterate` and  `Search`
+// that are implemented through a series of simple move methods (`down`,
+// `up`, `Right`) that modify its `path`. The `down` move method is the
+// analogous of the recursive call and the one in charge of visiting
+// the node (through the `VisitHandler`) and performing some user-defined
+// logic.
+//
+// The position of the `Traversal` is defined through a `path` (slice)
+// of nodes that form a parent-child relationship from the root to the
+// current node being visited (at a depth marked by the `level`). Each
+// of those nodes keep track of the link index its child in the path
+// belongs to in order for the traversal algorithms to know not only
+// what it the current nodes but which nodes (children) have already
+// been visited (assuming always a left to right, zero to `len(Links())`,
+// visiting order).
+//
+// TODO: Add a logic similar to `precalcNextBuf` to request several child
+// nodes in advanced as node promises.
+//
+// TODO: Revisit the name. `Traverser`? (is "traverser" the correct noun?),
+// `Iterator` (this would put to much emphasis on iterate whereas other
+// traversal operations like search are supported), `Topology`?
```

Comment:
`Walk` and `WalkFunc` are the terms chosen in some comparable standard library types, fwiw: https://golang.org/pkg/path/filepath/#Walk

Reply:
FWIW, “traverse/traversal/traverser” is a technically correct (and not uncommon) English term for this that I have seen in many libraries (it even has a [Wikipedia page](https://en.wikipedia.org/wiki/Tree_traversal)). That said, I think I’ve seen “walk/walking/walker” used just as often and it’s shorter :)